src/WebJobs.Extensions.SendGrid/Config/SendGridConfiguration.cs

```diff
@@ -38,12 +50,58 @@ public SendGridConfiguration()
         /// jobs are executed. In this case, job functions can specify minimal info in
         /// their bindings, for example just a Subject and Text body.
         /// </remarks>
-        public MailAddress ToAddress { get; set; }
+        public Email ToAddress { get; set; }
 
         /// <summary>
         /// Gets or sets the default "from" address that will be used for messages.
         /// This value can be overridden by job functions.
         /// </summary>
-        public MailAddress FromAddress { get; set; }
+        public Email FromAddress { get; set; }
+
+        internal ISendGridClientFactory ClientFactory { get; set; }
+
+        /// <inheritdoc />
+        public void Initialize(ExtensionConfigContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            INameResolver nameResolver = context.Config.NameResolver;
+            _defaultApiKey = nameResolver.Resolve(AzureWebJobsSendGridApiKeyName);
+
+            IConverterManager converterManager = context.Config.GetService<IConverterManager>();
+            converterManager.AddConverter<string, Mail>(SendGridHelpers.CreateMessage);
+            converterManager.AddConverter<JObject, Mail>(SendGridHelpers.CreateMessage);
+
+            BindingFactory factory = new BindingFactory(nameResolver, converterManager);
+            IBindingProvider outputProvider = factory.BindToAsyncCollector<SendGridAttribute, Mail>((attr) =>
+            {
+                string apiKey = FirstOrDefault(attr.ApiKey, ApiKey, _defaultApiKey);
+                ISendGridClient sendGrid = _sendGridClientCache.GetOrAdd(apiKey, a => ClientFactory.Create(a));
+                return new SendGridMessageAsyncCollector(this, attr, sendGrid);
+            });
+
+            IExtensionRegistry extensions = context.Config.GetService<IExtensionRegistry>();
+            extensions.RegisterBindingRules<SendGridAttribute>(ValidateBinding, nameResolver, outputProvider);
+        }
+
+        private void ValidateBinding(SendGridAttribute attribute, Type type)
+        {
+            string apiKey = FirstOrDefault(attribute.ApiKey, ApiKey, _defaultApiKey);
+
+            if (string.IsNullOrEmpty(apiKey))
+            {
+                throw new InvalidOperationException(
+                    string.Format("The SendGrid ApiKey must be set either via a '{0}' app setting, via a '{0}' environment variable, or directly in code via SendGridConfiguration.ApiKey.",
```

Comment:
Should this error message also mention that it can be set via SendGridAttribute.ApiKey?